### PR TITLE
Upgrade to python3

### PR DESCRIPTION
**Pull request by Josue-T**
## Problem
- Python 2 is depreciated
- Using the same uwsgi systemd unit for all appliction is not a good practice.

## Solution
- Migrate to python3
- Use the uwsgi helper to separate all applications

## PR Status
- [x] Code finished.
- [x] Tested with Package_check.
- [x] Fix or enhancement tested.
- [x] Upgrade from last version tested.
- [x] Can be reviewed and tested.

## Validation
---
*Minor decision*
- **Upgrade previous version** : 
- [ ] **Code review** : 
- [ ] **Approval (LGTM)** : 
- [ ] **Approval (LGTM)** : 
- **CI succeeded** : 
[Build Status](https://ci-apps-hq.yunohost.org/jenkins/job/searx_ynh%20PR61/)  

When the PR is marked as ready to merge, you have to wait for 3 days before really merging it.

### CODE REVIEW

File: scripts/upgrade [hunk 1]
```diff
@@ -123,11 +123,12 @@ ynh_system_user_create --username=$app
...
 
-rm -r $final_path/lib/python2.7/site-packages/setuptools $final_path/lib/python2.7/site-packages/setuptools-*
-virtualenv --system-site-packages "$final_path"
+# Clean venv if it still on python2
+test -e $final_path/bin/python3 || ynh_secure_remove --file=$final_path
+python3 -m venv --system-site-packages "$final_path"
 set +u; source $final_path/bin/activate; set -u
-pip install -U setuptools
-pip install --requirement $final_path/requirements-ynh.txt --upgrade
+pip3 install -U pip setuptools
```
Comment (maniackcrudelis) [inline scripts/upgrade:130]: To not install pip with pip was on purpose
https://github.com/YunoHost-Apps/searx_ynh/pull/31

**Comment by maniackcrudelis** [comment]
I do not see how this will fix #60
But anyway, to migrate to python3 is a good thing for this app.

However, last time it didn't work. And for now, It doesn't...

**Comment by Josue-T** [comment]
> However, last time it didn't work. And for now, It doesn't...

Well only the upgrade from the old version fail. Everything else work...

I'll fix this

**Comment by maniackcrudelis** [comment]
> Well only the upgrade from the old version fail. Everything else work...

And #60 is "Upgrade failed", but to says that the upgrade is a problem is obviously a minor problem

File: scripts/upgrade [hunk 2]
```diff
@@ -124,10 +124,16 @@ ynh_system_user_create --username=$app
 ynh_script_progression --message="Upgrading Searx..." --weight=7
 
 # Clean venv if it still on python2
-test -e $final_path/bin/python3 || ynh_secure_remove --file=$final_path
+test -e $final_path/bin/python3 || (
+    ynh_secure_remove --file=$final_path/lib/python*
```
Comment (maniackcrudelis) [inline scripts/upgrade:128]: Sure the wildcard work correctly through `ynh_secure_remove` ?

**Comment by Josue-T** [comment]
Yes if there are only one file which match.

**Comment by maniackcrudelis** [comment]
> Yes if there are only one file which match.

I do not really feel relieved...
It does not sound like a resilient way to remove the files !

I don't like that rm -r, which may be a rm only if there's only one file.
But to use the helper knowing that it doesn't work correctly with regex. That doesn't sound like a good idea...

**Comment by Josue-T** [comment]
> > Yes if there are only one file which match.
> 
> I do not really feel relieved...
> It does not sound like a resilient way to remove the files !
> 
> I don't like that rm -r, which may be a rm only if there's only one file.
> But to use the helper knowing that it doesn't work correctly with regex. That doesn't sound like a good idea...

In this case we need to use something like `rm -r $final_path/lib/python* || true`

**Comment by maniackcrudelis** [comment]
You mean something to ignore an error if the file is nonexistent ?
Like something that would "ignore nonexistent files" ?

So... `-f`

**Comment by Josue-T** [comment]
> You mean something to ignore an error if the file is nonexistent ?
> Like something that would "ignore nonexistent files" ?
> 
> So... `-f`

As you want

**Comment by maniackcrudelis** [comment]
If I can find some extra time, a block hole would be useful, I'll have a look to include a way to manage regex into that helper.

**Comment by maniackcrudelis** [comment]
https://github.com/YunoHost-Apps/Experimental_helpers/blob/master/ynh_secure_remove_regex/ynh_secure_remove_regex

**Comment by maniackcrudelis** [comment]
To use that helper with the same name into the _common.sh won't work, as the helper is overwritten by `source /usr/share/yunohost/helpers`

File: scripts/upgrade [hunk 3]
```diff
@@ -6,8 +6,8 @@
 # IMPORT GENERIC HELPERS
 #=================================================
 
-source _common.sh
 source /usr/share/yunohost/helpers
+source _common.sh
```
Comment (maniackcrudelis) [inline scripts/upgrade:10]: Seriously !?

I mean, dude you're supposed to be in the Apps group !

At least just say that you don't know or you don't want to do it, we'll do it. But come on dude...

**Comment by Thatoo** [comment]
could this update solve my issue https://github.com/YunoHost-Apps/searx_ynh/issues/62 ?

**Comment by maniackcrudelis** [comment]
Apparently last time we've tested this branch, the upgrade from a previous version wasn't working.
So I don't think it would be a good idea for now to try it.

**Pull request merged**